### PR TITLE
Teach `#isolation` to respect the flow-sensitive nature of actor initializers

### PR DESCRIPTION
**Pull request by DougGregor**
Actor initializers have a flow-sensitive property where they are isolated to the actor being initialized only after the actor instance itself is fully-initialized. However, this behavior was not being reflected in the expansion of `#isolation`, which was always expanding to `self`, even before `self` is fully formed.

This led to a source compatibility issue with code that used the async for..in loop within an actor initializer *prior* to the point where the actor was fully initialized, because the type checker is introducing the `#isolation` (SE-0421) but Definite Initialization properly rejects the use of `self` before it is initialized.

Address this issue by delaying the expansion of `#isolation` until after the actor is fully initialized. In SILGen, we introduce a new builtin for this case (and *just* this case) called `flowSensitiveSelfIsolation`, which takes in `self` as its argument and produces an `(any Actor)?`. Definite initialization does not treat this as a use of `self`. Rather, it tracks these builtins and replaces them either with `self` (if it is fully-initialized at this point) or `nil` (if it is not fully-initialized at this point), mirroring the flow-sensitive isolation semantics described in SE-0327.

Fixes rdar://127080037.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: include/swift/AST/Builtins.def [hunk 1]
```diff
@@ -899,14 +899,25 @@ BUILTIN_MISC_OPERATION(StartAsyncLetWithLocalBuffer, "startAsyncLetWithLocalBuff
 /// This is only supported under the task-to-thread concurrency model.
 BUILTIN_MISC_OPERATION(TaskRunInline, "taskRunInline", "", Special)
 
-/// flowSensitiveSelfIsolation<T: Actor>(_ actor: T) -> T?
+/// flowSensitiveSelfIsolation<T: Actor>(_ actor: T) -> (any Actor)?
```
Comment (ktoso) [inline include/swift/AST/Builtins.def:902]: 👍

**Review by ktoso — APPROVED** [review]
Great lgtm, Thank you!

**Pull request merged**